### PR TITLE
feat: standalone grandpa ranger

### PR DESCRIPTION
**Pull request by petscheit**
# Summary of changes
creates a standalone grandpa ranger application:
- gracefully handles any error that can pop up
- hot-swap ws connections on disconnect or error
- provides prometheus compatible /metrics & /status endpoint for monitoring

Reimplements the portal rpc's to fetch current gateway height

# Acceptance Checklist

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

## Dependencies

- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] Any teams which manage the dependencies have been notified of the breaking changes

## Bug (non-breaking change which fixes an issue):

- [ ] Have you **_written tests_** to protect against future occurrences of the bug?

## Feature (non-breaking change which adds functionality)

- [x] Have you **_seen it working_** in a development environment?
- [ ] Have you **_written tests_** for **_happy_** and **_unhappy_** paths?
- [x] Have you implemented this feature as per the **_issue acceptance criteria_**?

### Substrate

If this change involves substrate, have you remembered:

- [ ] Storage Migration?
- [x] RPC methods?
- [ ] Chainspec, both JSON specs & the module?
- [ ] Runtime versioning?
- [ ] Benchmarks?

## Breaking change (a feature that would cause existing functionality not to work as expected)

- [ ] Is the breaking change **_documented_**?
- [ ] Are your commits fully representative of the change?
- [ ] Has any previous functionality been **_deprecated_** and versioned?

## CI/CD

- [ ] If introducing new actions, have you **_looked at the action code_** for anything spurious?
- [ ] Have you written **_custom scripts_** for things that could be actions already on the marketplace?
- [ ] If introducing new actions, have you **_pegged a static version_**?

## Documentation Update

- [ ] Have you checked other documentation, such as the docs repository?
- [ ] If updating script documentation, have you **_tested it on both environments_**?

## Further comments

Feel free to explain in further detail your choices if this is a significant change.

### CODE REVIEW

File: client/packages/grandpa-ranger/config/local.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	rangeBreak: number,
+	targetGatewayId: string
+	circuitSigner: string,
+}
+
+export default {
+	circuit: {
+		rpc1: {
+			ws: "ws://localhost:9944",
+			http: "http://localhost:9933"
```
Comment (MaciejBaj) [inline client/packages/grandpa-ranger/config/local.ts:26]: WS_CIRCUIT_ENDPOINT flag

File: client/packages/grandpa-ranger/src/collect.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+			}
+
+		} catch (error) {
+			return reject(error)
+		}
+	})
+}
+
+// TODO: add types
+interface TransactionArguments {
```
Comment (palozano) [inline client/packages/grandpa-ranger/src/collect.ts:30]: Can these ones be more specific inside the interface?

File: client/packages/grandpa-ranger/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+class GrandpaRanger {
+	circuit: Connection;
+	target: Connection;
+	config: any;
+	prometheus: Prometheus;
+
+	constructor(config: any) {
+		this.config = config;
+		this.prometheus = new Prometheus();
+		this.prometheus.rangeBreak.inc(this.config.rangeBreak);
```
Comment (3h4x) [inline client/packages/grandpa-ranger/src/index.ts:17]: maybe rangeInterval?

File: client/packages/grandpa-ranger/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+		if(batches.length > 0) {
+			// calculate the total number of elements in the batches elements
+			const totalElements = batches.reduce((acc, curr) => acc + curr.range.length, 0)
+
+			this.submitToCircuit(batches)
+				.then((res) => {
+					console.log({"status": "Submitted", "range_size": totalElements, "circuit_block": res})
+					this.prometheus.nextSubmission.set(Date.now() + this.config.rangeBreak * 1000);
+					this.prometheus.successes.inc({rangeSize: totalElements, circuitBlock: res, timestamp: Date.now()})
```
Comment (3h4x) [inline client/packages/grandpa-ranger/src/index.ts:52]: timestamp, circuitblock and rangesize shouldnt be used as labels
https://prometheus.io/docs/practices/instrumentation/#do-not-overuse-labels

circuitblock should be different metric if its not the same as circuitHeight

File: client/packages/grandpa-ranger/src/prometheus.ts [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+		this.successes = new client.Counter(
+			{
+				name: 'successes',
+				help: 'Information on the latest successful submissions',
+				registers: [this.register],
+				labelNames: ['rangeSize', 'timestamp', 'circuitBlock']
+			}
+		);
+
+		this.successCount = new client.Counter(
```
Comment (3h4x) [inline client/packages/grandpa-ranger/src/prometheus.ts:63]: successes_total

https://prometheus.io/docs/practices/naming/#metric-names

File: client/packages/grandpa-ranger/src/prometheus.ts [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+		this.nextSubmission = new client.Gauge(
+			{
+				name: 'next_submission',
+				help: 'Unix timestamp of the next scheduled submission',
+				registers: [this.register]
+			}
+		)
+
+		this.successes = new client.Counter(
```
Comment (3h4x) [inline client/packages/grandpa-ranger/src/prometheus.ts:54]: i dont think this one will be useful, success_total will suffice

File: client/packages/grandpa-ranger/src/prometheus.ts [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+		this.successCount = new client.Counter(
+			{
+				name: 'success_counter',
+				help: 'Number of successful submissions',
+				registers: [this.register],
+			}
+		)
+
+		this.errors = new client.Counter(
```
Comment (3h4x) [inline client/packages/grandpa-ranger/src/prometheus.ts:71]: same as successes

File: client/packages/grandpa-ranger/src/prometheus.ts [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+			{
+				name: 'errors',
+				help: 'Information on the latest errored submissions',
+				registers: [this.register],
+				labelNames: ['rangeSize', 'timestamp']
+			}
+		);
+
+		this.errorCount = new client.Counter(
```
Comment (3h4x) [inline client/packages/grandpa-ranger/src/prometheus.ts:80]: errors_total

File: client/packages/grandpa-ranger/src/prometheus.ts [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+			{
+				name: 'error_counter',
+				help: 'Number of errored submissions',
+				registers: [this.register],
+			}
+		)
+
+		this.circuitDisconnected = new client.Counter(
```
Comment (3h4x) [inline client/packages/grandpa-ranger/src/prometheus.ts:88]: timestamp shouldnt be in label, its already internal part of this metric

File: client/packages/grandpa-ranger/src/prometheus.ts [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+		this.rangeBreak = new client.Counter({
+			name: 'range_loop',
+			help: 'The number of seconds between each range submission',
+			registers: [this.register]
+		})
+
+		this.startServer()
+	}
+
+	startServer() {
```
Comment (3h4x) [inline client/packages/grandpa-ranger/src/prometheus.ts:131]: its nice to have metric `up` when server starts with value 1 when its running

File: client/packages/grandpa-ranger/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+		if(range.length > 10) {
+			range = range.slice(0, 10);
+		}
+		return new Promise(async (resolve, reject) => {
+			try {
+				if(this.circuit.sdk && this.circuit.isActive) {
+					let tx = this.circuit.sdk.circuit.tx.createBatch(range.map(args => {
+						let submit;
+						// select the correct submit function based on the targetGatewayId
+						if(this.config.targetGatewayId === "roco") {
```
Comment (3h4x) [inline client/packages/grandpa-ranger/src/index.ts:81]: targetGatewayId should be a label that we add to all metrics

**Review by 3h4x — CHANGES_REQUESTED** [review]
nice nice.
I left some comments about prometheus metrics

**Review by 3h4x — CHANGES_REQUESTED** [review]
nice nice.
I left some comments about prometheus metrics

File: client/packages/grandpa-ranger/src/index.ts (hunk 3 above)
Comment (petscheit) [reply client/packages/grandpa-ranger/src/index.ts:17]: much better. My creativity for variable naming is not great haha

File: docs/main/docs/collator/testnet/testnet-collator.md [hunk 12]
```diff
@@ -5,7 +5,7 @@ sidebar_position: 1
 # Run a t0rn Testnet Collator
 
 
-This guide outlines the essential minimum of steps required to run a collator for t0rn - a release candidate of t3rn on the Rococo testnet. This guide uses the `v1.15.1-rc.0` release, however always use the latest available version.
+This guide outlines the essential minimum of steps required to run a collator for t0rn - a release candidate of t3rn on the Rococo testnet. This guide uses the `v1.16.1-rc.0` release, however always use the latest available version.
```
Comment (chiefbiiko) [inline docs/main/docs/collator/testnet/testnet-collator.md:8]: Would separate the collator version updates in our docs from this PR as its already out of date again. This should ideally be handled as part of the collator release.

**Review by 3h4x — APPROVED** [review]
🚀

File: docs/main/docs/collator/testnet/testnet-collator.md (hunk 12 above)
Comment (3h4x) [reply docs/main/docs/collator/testnet/testnet-collator.md:8]: this is merge conflict. those changes are automated with cargo release during merge to development

**Review by MaciejBaj — APPROVED** [review]

**Pull request merged**